Src/VimCore/EditorUtil.fs

```diff
@@ -254,49 +296,140 @@ type SnapshotLineRange  =
 /// This is the representation of a point within a particular line.  It's common
 /// to represent a column in vim and using a SnapshotPoint isn't always the best
 /// representation.  Finding the containing ITextSnapshotLine for a given 
-/// SnapshotPoint is an allocating operation and often shows up as a critical 
-/// metric in profiling.  This structure pairs the two together in a type safe fashion
+/// SnapshotPoint is an allocating operation and often shows up as a critical
+/// metric in profiling.  This structure pairs the two together in a type safe fashion.
+///
+/// Conceptually, a "column" corresponds to a single logical character:
+/// - a normal character
+/// - a line break (which may consist of one or more physical characters)
+/// - a UTF32 character (represented by two UTF16 characters called the "surrogate pair")
+/// Alternatively, a "column" represents the places where it is valid to set the caret.
 [<Struct>]
 [<NoEquality>]
 [<NoComparison>]
-type SnapshotColumn 
-    (
-        _snapshotLine: ITextSnapshotLine,
-        _column: int
-    ) =
+type SnapshotColumn =
 
-    new (point: SnapshotPoint) = 
-        let line = point.GetContainingLine()
-        let column = point.Position - line.Start.Position
-        SnapshotColumn(line, column)
+    val private _line: ITextSnapshotLine
+    val private _positions: int array
+    val private _columnNumber: int
 
-    member x.IsStartOfLine = _column = 0
-
-    member x.IsInsideLineBreak = EditorCoreUtil.IsInsideLineBreak x.Point x.Line
-
-    member x.Line = _snapshotLine
-
-    member x.LineNumber = _snapshotLine.LineNumber
-
-    member x.Snapshot = _snapshotLine.Snapshot
-
-    member x.Point = _snapshotLine.Start.Add(_column)
-
-    member x.Column = _column
-
-    member x.Add count = 
-        let column = _column + count
-        if column > 0 && column < _snapshotLine.LengthIncludingLineBreak then
-            SnapshotColumn(_snapshotLine, _column + count)
-        else
-            let point = x.Point.Add count
-            SnapshotColumn(point)
+    /// Constructor for a column corresponding to a snapshot point
+    /// that "parses" the whole containing line and builds a lookup
+    // table of physical positions corresponding to logical columns
+    new (point: SnapshotPoint) =

```

Comment:
Part of the motivation for the old constructor taking an `ITextSnapshotLine` was performance. Every call to `point.GetContainingLine` produces a new instance of `ITextSnapshotLine`. Using a `SnapshotPoint` based ctor meant that iterating every point in an `ITextSnapshot` allocated a new `ITextSnapshotLine` instance. That was actually a problem and traces lead me to the current form of taing an `ITextSnapshotLine`. Forced the caller to be a bit smarter about when the allocation should occur.

Reply:
The new API retains all of the performance advantages of the old API. The old API also had a constructor that just supplies a snapshot point. However, the new API has a constructor that takes a `SnapshotColumn` *instead* of an `ITextSnapshotLine` because not only do we need to carry over the snapshot line (like the old API), we also need to carry over the computed positions table. The positions table is an array and we want to avoid its allocation as well.

But believe me, I was extremely careful to avoid introducing any unnecessary inefficiency. Unfortunately, because our text buffer is effectively encoded, it is impossible to make a surgical change and have all of VsVim "just work."

Reply:
I improved the efficiency of the new API even more!